assets/src/setup/components/unsaved-changes-warning/index.js

```diff
@@ -0,0 +1,37 @@
+/**
+ * WordPress dependencies
+ */
+import { useEffect, useContext, useCallback } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import { Options } from '../options-context-provider';
+
+export function WizardUnsavedChangesWarning() {
+	const { hasChanges, hasSaved } = useContext( Options );
+
+	/**
+	 * Warns the user if there are unsaved changes before leaving the wizard.
+	 *
+	 * @param {Event} event `beforeunload` event.
+	 */
+	const warnIfUnsavedChanges = useCallback( ( event ) => {
+		if ( hasChanges && ! hasSaved ) {
+			event.returnValue = __( 'This page has unsaved changes. Are you sure you want to leave?', 'amp' );
+		}
+
+		return null;
+	}, [ hasChanges, hasSaved ] );
+
+	useEffect( () => {
+		window.addEventListener( 'beforeunload', warnIfUnsavedChanges );
```

Comment:
It is preferable to only add the `beforeunload` handler once the app has actually entered the dirty state (`hasChanges && ! hasSaved`). The reason for this is [explained on MDN](https://developer.mozilla.org/en-US/docs/Web/API/Window/beforeunload_event):

<blockquote><p>Attaching an event handler/listener to <code>window</code> or <code>document</code>'s <code>beforeunload</code> event prevents browsers from using in-memory page navigation caches, like <a href="/en-US/docs/Mozilla/Firefox/Releases/1.5/Using_Firefox_1.5_caching">Firefox's Back-Forward cache</a> or <a class="external" href="https://webkit.org/blog/516/webkit-page-cache-ii-the-unload-event/" rel="noopener">WebKit's Page Cache</a>.</p></blockquote>